### PR TITLE
Improve stability by reducing dependency on dollar

### PR DESCRIPTION
**Pull request by ivogabe**
**Description**
This pull request reduces the dependency of our codebase on dollar (`$`), by replacing it with `id`.

Dollar is actually an instantiation of the identity function, specialized to functions. Hence it makes more sense to use the more general `id`. Furthermore, `id` is an eta reduction of `$`.

I couldn't remove all uses of `$`, but this should already improve the situation a lot.

**Motivation and context**
Any instabilities of the dollar would impact Accelerate. If you for instance compare the dollar with the Venezuelan Bolivar, you'll see that dollar is actually quite unstable.

**How has this been tested?**
The nofib tests still pass.

**Types of changes**
What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Checklist**
Go over all the following points, and put an `x` in all the boxes that apply. If you're unsure about any of these, don't hesitate to ask. We're here to help!

- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by robbert-vdh** [comment]
Did you consider replacing the dollar with a new, local definition:

```haskell
{-# INLINE (€) #-}
(€) :: forall r a (b :: TYPE r). (a -> b) -> a -> b
f € x =  f x

infixr 0 €
```

You could also try replacing it with `₿`, but that might make it even more unstable.

File: src/Data/Array/Accelerate/AST.hs [hunk 1]
```diff
@@ -1247,11 +1247,11 @@ liftMessage aR (Message _ fmt msg) =
       -- back to displaying in representation format
       fmtR :: ArraysR arrs' -> Q (TExp (arrs' -> String))
       fmtR TupRunit                         = [|| \() -> "()" ||]
-      fmtR (TupRsingle (ArrayR ShapeRz eR)) = [|| \as -> showElt $$(liftTypeR eR) $ linearIndexArray $$(liftTypeR eR) as 0 ||]
+      fmtR (TupRsingle (ArrayR ShapeRz eR)) = [|| \as -> showElt $$(liftTypeR eR) `id` linearIndexArray $$(liftTypeR eR) as 0 ||]
       fmtR (TupRsingle (ArrayR shR eR))     = [|| \as -> showArray (showsElt $$(liftTypeR eR)) (ArrayR $$(liftShapeR shR) $$(liftTypeR eR)) as ||]
       fmtR aR'                              = [|| \as -> showArrays $$(liftArraysR aR') as ||]
```
Comment (tomsmeding) [inline src/Data/Array/Accelerate/AST.hs:1252]: It's kind of unfortunate that TemplateHaskell syntax itself is not only dependent on the currency in question, but also on large amounts of it, right?

**Comment by ivogabe** [comment]
Maybe we should do that for the occurrences of dollar which I couldn't remove, as the precedence and associativity of id is different. Euro also isn't fully stable, but that should work if we only use it a few times. Though I would prefer to define € pointfree:
```
(€) = id
```

File: src/Data/Array/Accelerate/AST.hs (hunk 1 above)
Comment (ivogabe) [reply src/Data/Array/Accelerate/AST.hs:1252]: That might even be a larger problem!

Comment (ivogabe) [reply src/Data/Array/Accelerate/AST.hs:1252]: Though that would require a GHC plugin to fix it probably

**Comment by dpvanbalen** [comment]
I have some doubts regarding the 'checklist'

File: src/Data/Array/Accelerate/Error.hs [hunk 2]
```diff
@@ -33,13 +33,13 @@ data Check = Bounds | Unsafe | Internal
 -- | Issue an internal error message
 --
 internalError :: HasCallStack => String -> a
-internalError = withFrozenCallStack $ error Internal
+internalError = withFrozenCallStack `id` error Internal
```
Comment (tomsmeding) [inline src/Data/Array/Accelerate/Error.hs:36]: Doesn't compile:

```
.../src/Data/Array/Accelerate/Error.hs:36:17: error:
    • Couldn't match type ‘?callStack::CallStack’ with ‘[Char]’
      Expected type: (String -> String -> a) -> String -> a
        Actual type: (HasCallStack => String -> a) -> String -> a
    • In the first argument of ‘id’, namely ‘withFrozenCallStack’
      In the expression: withFrozenCallStack `id` error Internal
      In an equation for ‘internalError’:
          internalError = withFrozenCallStack `id` error Internal
   |
36 | internalError = withFrozenCallStack `id` error Internal
   | 
```

Comment (ivogabe) [reply src/Data/Array/Accelerate/Error.hs:36]: Oh is that because of the special handling of `($)` in GHC?

Comment (tomsmeding) [reply src/Data/Array/Accelerate/Error.hs:36]: Yes, `($)` is levity-polymorphic. I'm working on a fix.

**Comment by tomsmeding** [comment]
I pushed a further improvement; this pattern should allow removing all occurrences of `($)`, though I only applied the relevant changes in `AST.hs` for now. Applying this across the entire codebase can be entirely mechanical; adjusting the imports is probably the hardest part to do automatically.

However, the major problem that I don't see a solution to is TemplateHaskell's [fixation on money](https://cdn4.iconfinder.com/data/icons/smiley-2-1/32/190-512.png). But I guess if one just avoids using `runQ`, one can avoid using the TemplateHaskell code in end-user programs. Perhaps this is good enough.

**Comment by ivogabe** [comment]
But now you eta-expanded id! And it was actually the goal to use the generic identity function, not specialized to functions!

**Comment by tomsmeding** [comment]
@ivogabe Oops! It seems the definition can be eta-reduced successfully:

```haskell
import qualified Prelude

id :: forall r a (b :: TYPE r). (a -> b) -> a -> b
id = Prelude.id
```

but attempting to un-specialise the function to a generic identity function fails:

```haskell
id :: forall r (a :: TYPE r). a -> a
id x = x

-- .../src/Data/Array/Accelerate/Uncurrency.hs:16:4: error:
--     A levity-polymorphic type is not allowed here:
--       Type: a
--       Kind: TYPE r
--     In the type of binder ‘x’
--    |
-- 16 | id x = x
--    |    ^
```

I'm not sure if the perfect definition is attainable.

**Comment by ivogabe** [comment]
Could we also use something like parentheses on the places where the normal id doesn't work?

**Comment by tomsmeding** [comment]
@ivogabe Surely you jest, it's well known that excessive usage of parentheses [led to the AI winter](https://en.wikipedia.org/wiki/AI_winter#The_collapse_of_the_LISP_machine_market) &mdash; that's dangerous ground.

**Comment by ivogabe** [comment]
I'm afraid this controversial PR won't make it :(. So let's just pretend this was an April fools joke.

**Pull request closed without merging**